### ISSUE TITLE
Enhancement Add async support

### ISSUE THREAD

**Issue by tomekmarchi** [ractivejs/ractive#3062]
Add async/await support for all initialization functions for example data(if a function is given), onrender etc. This would allow async dependencies or operations to occur that may be relevant to the later data. For the large projects I find myself involved in this is a huge asset. Code becomes easier to read, far more condensed and introduces flow control for async operations. It seems all that would need to be done is have an await set for the possible async ractive methods.

```
{
 async data(){
   const test = await asyncOp();
   return {
    test
   };
 }
}

```

**Comment by ghost** [comment]
TL;DR: There's probably a better way that making everything async 😄 

Several topics of the same idea were brought up before on different parts of Ractive:

- https://github.com/ractivejs/ractive/issues/2075
- https://github.com/ractivejs/ractive/issues/2130
- https://github.com/ractivejs/ractive/issues/2512
- https://github.com/ractivejs/ractive/issues/632

If you want to load data asynchronously, [we have Promise adaptors](https://ractive.js.org/integrations/#adaptors), two in fact (might be dated, but they're not that complex to just copy and modify either).

I'm also using Ractive ala Flux/Redux so I might be a bit biased. In this case, my components don't care how or when the data loads because that bit is external. I see `data` as more of "default values"/"initial data structure" rather than housing the actual loading mechanism. Baking this into Ractive might be overkill.

If ever this were to be implemented, I'm not sure how it would affect initialization. Like setting up the mappings across component boundaries, dependencies, resolutions and etc. because on initialization, you don't have the data structure. You only have a promise that does not guarantee what comes back, or if it comes back at all.

On a related note, Ractive chose to be synchronous for the most part (except for the parts that can't be synchronous) so that it's easy to debug. You don't have to deal with operations that are disconnected from their sources, stack traces that originate from strange places, and that sort of stuff.

**Comment by evs-chris** [comment]
I would quite like to have full async support both in-template and for the API, but given the synchronous nature of ractive's guts, I haven't the foggiest how to implement either in the immediate future. I think fixing the adaptor API would be a good first step, as we could probably get functionality in place that would allow the root model to be adapted - which would allow a promise-based root model that could update upon resolution.

**Comment by tomekmarchi** [comment]
I can handle it boys. Don't worry gentlemen I am on the case I shall bring a TARR1 agonist with me. Already have a working implementation and it's heavenly. @evs-chris <3

**Comment by tomekmarchi** [comment]
I'll shoot chris over the details so I can avoid the pull-request. However, I'm going to submit a fully async by default one as that is exactly where components and the entire JS echosystem is heading.

**Comment by tomekmarchi** [comment]
On the debug point. With increasing support for Async and it's fast maturity debugging has become very easy to work with at this rate. The issues that stand today will not be here for the world of tomorrow. Therefore, I am highly supportive of future proof tech. Hence my full support for this move.

**Comment by tomekmarchi** [comment]
Overkill no, necessary most definitely. The way I see it is pretty simple at this point. Evolve with the times or die. If there is opposition to this I'll just fork it and support the Async variety. As at this point when dealing with highly reactive components with real-time data for multiple industrial applications I absolutely need this. That's just the honest truth gentlemen. None the less I will make the changes, describe the modifications and send it over. God speed boys.

**Comment by PaulMaly** [comment]
@tomekmarchi maybe it could be closed by [#3171](https://github.com/ractivejs/ractive/pull/3171) ?
